Source/Pawnmorphs/Esoteria/UserInterface/Settings/Dialog_BlacklistAnimal.cs

```diff
@@ -0,0 +1,177 @@
+			_animalListBox.Draw(inRect, 0, curY, totalHeight, (item, listing) =>
+			{
+				bool current = _blacklistedAnimalDictonary.TryGetValue(item, false);
+				listing.CheckboxLabeled(item.LabelCap, ref current, item.modContentPack.ModMetaData.Name);
+				_blacklistedAnimalDictonary[item] = current;
+			});
+
+			// Draw the apply, reset and cancel buttons.
+			if (Widgets.ButtonImage(new Rect(5, inRect.height - 33, RAW_BUTTON_SIZE.x, RAW_BUTTON_SIZE.y), TexButton.Rename))
+			{
+				ShowRawInput();
+			}
+
+			float buttonVertPos = inRect.height - Math.Max(APPLY_BUTTON_SIZE.y, Math.Max(RESET_BUTTON_SIZE.y, CANCEL_BUTTON_SIZE.y));
+			float applyHorPos = inRect.width / 2 - APPLY_BUTTON_SIZE.x - RESET_BUTTON_SIZE.x / 2 - SPACER_SIZE;
+			float resetHorPos = inRect.width / 2 - RESET_BUTTON_SIZE.x / 2;
+			float cancelHorPos = inRect.width / 2 + RESET_BUTTON_SIZE.x / 2 + SPACER_SIZE;
+			if (Widgets.ButtonText(new Rect(applyHorPos, buttonVertPos, APPLY_BUTTON_SIZE.x, APPLY_BUTTON_SIZE.y), APPLY_BUTTON_TEXT))
+			{
+				OnAcceptKeyPressed();
+			}
+			if (Widgets.ButtonText(new Rect(resetHorPos, buttonVertPos, RESET_BUTTON_SIZE.x, RESET_BUTTON_SIZE.y), RESET_BUTTON_TEXT))
+			{
+				RimWorld.SoundDefOf.Checkbox_TurnedOff.PlayOneShotOnCamera(null);
+				RefreshList();
+			}
+			if (Widgets.ButtonText(new Rect(cancelHorPos, buttonVertPos, CANCEL_BUTTON_SIZE.x, CANCEL_BUTTON_SIZE.y), CANCEL_BUTTON_TEXT))
+			{
+				OnCancelKeyPressed();
+			}
+		}
+
+
+		private void ShowRawInput()
+		{
+			List<string> exportCollection = new List<string>();
+			exportCollection.AddRange(_blacklistedAnimalDictonary.Where(x => x.Value == false).Select(x => x.Key.defName).ToList());
+			exportCollection = exportCollection.Distinct().ToList();
+
+			string stringValue = string.Join(",", exportCollection);
+			Dialog_Textbox textBox = new Dialog_Textbox(stringValue, true, new Vector2(300, 100));
+
+			textBox.ApplyAction = (value) =>
+			{
+				try
+				{
+					_settingsReference.Clear();
+					_settingsReference.AddRange(value.Split(',').Where(x => String.IsNullOrWhiteSpace(x) == false).Distinct());
+					RefreshList();
+				}
+				catch (Exception)
+				{
+					Find.WindowStack.Add(new Dialog_Popup(FAILED_PARSING_TEXT, new Vector2(300, 100)));
+				}
+			};
+
+			Find.WindowStack.Add(textBox);
+		}
+
+		private void ApplyChanges()
+		{
+			// We want to persist any definitions assigned for defnames that could not be loaded into the game at present time.
+			// Like animals from currently disabled mods.
+			// Take the old list, add all disabled animals (this will cause duplicates) and then apply distinct.
+			_settingsReference.AddRange(_blacklistedAnimalDictonary.Where(x => x.Value == false).Select(x => x.Key.defName).ToList());
+			_settingsReference = _settingsReference.Distinct().ToList();
+
+			foreach (var animal in _blacklistedAnimalDictonary)
+			{
```

Comment:
I have forgotten to add the black listing on initialization part. Oops!

This was just my attempt to _not_ require a full restart to ban additional animals.

The black list should only exist in config form and then parsed and applied to all the individual former human components.

Otherwise you get an increasing performance hit on testing former humans with each banned creature.

Reply:
You could convert the blacklist into a HashSet to ensure a more or less a constant-time check.  I don't think Scribe works with HashSets natively, but it wouldn't be too hard to just make an intermediate set on loading and after changing the blacklist.

Since the check only happens upon trying to generate former humans or turn humans into animals (which happens far less often than once a tick), the performance hit shouldn't be a big deal anyway.